designs/cloud_formation_api_gateway_support.md

```diff
@@ -0,0 +1,377 @@
+
+The AWS::ApiGateway::Stage will initially support the schema:
+
+```yaml
+Prod:
+Type: AWS::ApiGateway::Stage
+Properties:
+  StageName: Prod
+  Description: Prod Stage
+  Variables:
+    Stack: Prod
+  RestApiId: RestApi`
+```
+
+Since the stages can show up in any order as the code is iterating through the resources, There are two approaches to consider too quickly stitch the api and stages together. 
+
+*Approach #1:* 
+Arbitrarily pick one of the stages and use that on all the Api
+Pros: Similar to the current approach
+Cons: The customer may not be able to view their environments in different stages.
+
+*Approach: #2: *
+Create a dictionary of all the stages and display the url as localhost:3000/<Stage Name>/routeName
+Pros: This will allow people to view their code in different stages or environments.
+Cons: This will require refactoring parts of the SamApiProvider that handles the Api. Customers usually don't use different stages to test and it may not be worth the effort to test. 
+
+I recommend approach #1 as it follows the current standard and there is some data that very little people use the <Stage Name> when defining it. 
+
+**AWS::ApiGateway::Deployment **
+The deployment type defines which Apis objects should be available to the public. Normally, in the SAM template everything defined will have a deployment component, but the CloudFormation templates may have more bloat and information in them defining certain apis and deployments for Multi-stages.
+
+One approach is to support multiple stages with multiple deployments and the code will filter out the stages that are not deployed or defined. One problem with this approach is that the RestApiId, a characteristic of the RestApi, for a deployment is only localhost and a single endpoint and the current schema doesn't allow deploying on multiple ports and multiple local domains. This will bring in unneeded complexity such as requiring the customer defining N valid ports and N valid domains.
+
+Another approach, which I recommend, is to ignore the flag altogether as it always associated with a stage and to randomly pick a stage. 
+
+## **Updating the code**
+
+First, the Apis in AWS::ApiGateway::RestApi  will be parsed in the SAM Template in a similar way to the current AWS::Serverless:Api code. Stages and other Meta information in the Api will be parsed to support the CloudFormation template. The code will then be refactored to first translate the SAM templates inputted into the CloudFormation template using the SamTranslator. Once the Api code has been refactored, the function code can be refactored so that Serverless code is refactored out. 
```

Comment:
Not now but when you get to doing the Transformation to CFN, we can expand this doc or create a new one. I think there is enough interesting tradeoffs and decisions in that to warrant it.